src/Core/BaseKernel.php

```diff
@@ -0,0 +1,515 @@
+        if ($defaultSet instanceof ThemeManifest) {
+            $defaultSet->setProject(
+                ModuleManifest::config()->get('project')
+            );
+            $defaultSet->init(
+                $this->getIncludeTests(),
+                $flush,
+                $this->getIgnoredCIConfigs()
+            );
+        }
+    }
+
+    /**
+     * Include all _config.php files
+     */
+    protected function bootConfigs()
+    {
+        global $project;
+        $projectBefore = $project;
+        $config = ModuleManifest::config();
+        // After loading all other app manifests, include _config.php files
+        $this->getModuleLoader()->getManifest()->activateConfig();
+        if ($project && $project !== $projectBefore) {
+            Deprecation::notice('5.0', '$project global is deprecated');
+            $config->set('project', $project);
+        }
+    }
+
+    /**
+     * Turn on error handling
+     * @throws Exception
+     */
+    protected function bootErrorHandling()
+    {
+        // Register error handler
+        $errorHandler = Injector::inst()->get(ErrorHandler::class);
+        $errorHandler->start();
+
+        // Register error log file
+        $errorLog = Environment::getEnv('SS_ERROR_LOG');
+        if ($errorLog) {
+            $logger = Injector::inst()->get(LoggerInterface::class);
+            if ($logger instanceof Logger) {
+                $logger->pushHandler(new StreamHandler($this->basePath . '/' . $errorLog, Logger::WARNING));
+            } else {
+                user_error("SS_ERROR_LOG setting only works with Monolog, you are using another logger", E_USER_WARNING);
+            }
+        }
+    }
+
+
+
+    /**
+     * Get the environment type
+     *
+     * @return string
+     *
+     * @deprecated 5.0 use Director::get_environment_type() instead. Since 5.0 it should return only if kernel overrides. No checking SESSION or Environment.
+     */
+    public function getEnvironment()
+    {
+        // Check set
+        if ($this->enviroment) {
+            return $this->enviroment;
+        }
+
+        // Check saved session
+        $env = $this->sessionEnvironment();
+        if ($env) {
+            return $env;
+        }
+
+        // Check getenv
+        if ($env = Environment::getEnv('SS_ENVIRONMENT_TYPE')) {
+            return $env;
+        }
+
+        return self::LIVE;
+    }
+
+    /**
+     * Check or update any temporary environment specified in the session.
+     *
+     * @return null|string
+     *
+     * @deprecated 5.0 Use Director::get_session_environment_type() instead
+     */
+    protected function sessionEnvironment()
+    {
+        if (!$this->booted) {
+            // session is not initialyzed yet, neither is manifest
+            return null;
+        }
+
+        return Director::get_session_environment_type();
+    }
+
+    abstract public function boot($flush = false);
+
+    abstract public function isFlushed();
```

Comment:
As mentioned elsewhere, no need for this to be abstract, the implementation is the same in both CoreKernel + DatabaselessKernel, so may as well just do it here

Reply:
My instinct would be to keep it as is.

Looking at the implementation of `isFlushed`, it seems like it's tied in to the implementation of `boot`. Since `boot` is abstract, it would seem to make some sense to say that `isFlushed` should also be abstract. 

Otherwise, we're putting an implicit requirement on the sub classes to update the `isFlushed` variable in the boot method. Defining `isFlushed` as abstract makes that requirement explicit.